### PR TITLE
Add date to gamelog

### PR DESCRIPTION
**Pull request by lewri**
*Fixes #*

**Describe what the proposed change does**
- Append date to top of gamelog output
- This may be more useful in future, if we ever wanted to 'append' new gamelog entries to the file rather than overwriting it (though this could lead to bloat)

### CODE REVIEW

File: CorsixTH/Lua/app.lua [hunk 1]
```diff
@@ -1827,15 +1827,16 @@ end
...
   local compile_opts = TH.GetCompileOptions()
   local comp_details = {}
+  local date = os.date("%Y-%m-%d %H:%M:%S") -- Use ISO 8601 standard
   for key, value in pairs(compile_opts) do
     table.insert(comp_details, key .. ": " .. tostring(value))
   end
   table.sort(comp_details)
-  local compiled = string.format("Compiled with %s\nSDL renderer: %s\n",
-    table.concat(comp_details, ", "), self.video:getRendererDetails())
+  local compiled = string.format("Gamelog generated on %s\nCompiled with %s\nSDL renderer: %s\n",
```
Comment (lewri) [inline CorsixTH/Lua/app.lua:1835]: Now potentially out of scope of the function.. maybe be isolated to the gamelog functions instead (world:dumpgamelog and app:initgamelog).

**Comment by lewri** [comment]
Blocked currently by #2291 unless I integrate this PR into it.

**Comment by tobylane** [comment]
I like it, I have a minor suggestion to separate the date line for clarity.

````lua
function App:getSystemInfo()
  local date = string.format("Gamelog generated on %s\n", os.date("%Y-%m-%d %H:%M:%S")) -- Use ISO 8601 standard
  local compile_opts = TH.GetCompileOptions()
  local comp_details = {}
  for key, value in pairs(compile_opts) do
    table.insert(comp_details, key .. ": " .. tostring(value))
  end
  table.sort(comp_details)
  local compiled = string.format("Compiled with %s\nSDL renderer: %s\n",
      table.concat(comp_details, ", "), self.video:getRendererDetails())
  local running = string.format("%s run with api version: %s, game version: %s, savegame version: %s\n",
      compile_opts.jit or _VERSION, tostring(corsixth.require("api_version")),
      self:getVersion(), tostring(SAVEGAME_VERSION))
  return (date .. compiled .. running)
end
````

**Comment by lewri** [comment]
I'm not sure if Date should be part of scope of system information though? Do you think it should?

> I like it, I have a minor suggestion to separate the date line for clarity.
> 
> ```lua
> function App:getSystemInfo()
>   local date = string.format("Gamelog generated on %s\n", os.date("%Y-%m-%d %H:%M:%S")) -- Use ISO 8601 standard
>   local compile_opts = TH.GetCompileOptions()
>   local comp_details = {}
>   for key, value in pairs(compile_opts) do
>     table.insert(comp_details, key .. ": " .. tostring(value))
>   end
>   table.sort(comp_details)
>   local compiled = string.format("Compiled with %s\nSDL renderer: %s\n",
>       table.concat(comp_details, ", "), self.video:getRendererDetails())
>   local running = string.format("%s run with api version: %s, game version: %s, savegame version: %s\n",
>       compile_opts.jit or _VERSION, tostring(corsixth.require("api_version")),
>       self:getVersion(), tostring(SAVEGAME_VERSION))
>   return (date .. compiled .. running)
> end
> ```

At the moment I've changed it to this, but haven't pushed it yet.

```lua
function App:getSystemInfo()
  local compile_opts = TH.GetCompileOptions()
  local comp_details = {}
  local gen_date = os.date("%Y-%m-%d %H:%M:%S") -- Use ISO 8601 standard
  for key, value in pairs(compile_opts) do
    table.insert(comp_details, key .. ": " .. tostring(value))
  end
  table.sort(comp_details)
  local compiled = string.format("Compiled with %s\nSDL renderer: %s\n",
      table.concat(comp_details, ", "), self.video:getRendererDetails())
  local running = string.format("%s run with api version: %s, game version: %s, savegame version: %s\n",
      compile_opts.jit or _VERSION, tostring(corsixth.require("api_version")),
      self:getVersion(), tostring(SAVEGAME_VERSION))
  return gen_date, compiled .. running
end
```

File: CorsixTH/Lua/app.lua [hunk 2]
```diff
@@ -1827,15 +1827,16 @@ end
 function App:getSystemInfo()
   local compile_opts = TH.GetCompileOptions()
   local comp_details = {}
+  local date = os.date("%Y-%m-%d %H:%M:%S") -- Use ISO 8601 standard
```
Comment (TheCycoONE) [inline CorsixTH/Lua/app.lua:1830]: For what it's worth ISO 8501 requires a T not a ` `. Remove the misleading comment or fix the format. I have no strong preference.

Comment (lewri) [reply CorsixTH/Lua/app.lua:1855]: removed comment

File: CorsixTH/Lua/app.lua [hunk 3]
```diff
@@ -1849,19 +1851,23 @@ function App:isAudioEnabled()
...
   local compiled = string.format("Compiled with %s\nSDL renderer: %s\n",
-    table.concat(comp_details, ", "), self.video:getRendererDetails())
+      table.concat(comp_details, ", "), self.video:getRendererDetails())
   local running = string.format("%s run with api version: %s, game version: %s, savegame version: %s\n",
-    compile_opts.jit or _VERSION, tostring(corsixth.require("api_version")),
-    self:getVersion(), tostring(SAVEGAME_VERSION))
-  return (compiled .. running)
+      compile_opts.jit or _VERSION, tostring(corsixth.require("api_version")),
+      self:getVersion(), tostring(SAVEGAME_VERSION))
+  return gen_date, compiled .. running
```
Comment (TheCycoONE) [inline CorsixTH/Lua/app.lua:1870]: Not clear that we should be mixing conventions like this for this case.

What I mean is: getSystemInfo is currently concatenating the exact string to output in the game log. Now it's doing part of that construction and relying on the caller to do the rest. What set off alarms for me was that it was no longer providing it's primary function in the first return value, which is the normal convention.

I understand you're trying to abstract the function from it's use. I think we should do one of three things:
1. give up pretending this is an abstract function and call it `gameLogHeader()` and include `Gamelog generated on %s\n` in the returned string.
2. Include the date string in the returned string in such a way that it need not be specific to a game log e.g:
```
2022-12-07 08:59:00
Compiled with ...
```
3. Call os.date at both call sites where you're currently specifying `Gamelog generated on`.

3 feels like the worst option, the duplication you already have with the string is a good smell.

**Review by TheCycoONE — CHANGES_REQUESTED** [review]

File: CorsixTH/Lua/app.lua (hunk 3 above)
Comment (lewri) [reply CorsixTH/Lua/app.lua]: > 1. give up pretending this is an abstract function and call it `gameLogHeader()` and include `Gamelog generated on %s\n` in the returned string

Option 1 done

**Pull request merged**